inference-engine/src/mkldnn_plugin/nodes/ctc_greedy_decoder.cpp

```diff
@@ -0,0 +1,157 @@
+        const size_t BC = B * C;
+        const size_t CB1 = C * (B - 1);
+
+        const int blankIndex = C - 1;
+
+        std::vector<size_t> sequenceLengthB(B, 0);
+        parallel_for(B, [&](size_t b) {
+            for (size_t t = 0; t < T; t++) {
+                if (sequenceLengths[B * t + b] == 0.f)
+                    break;
+                sequenceLengthB[b]++;
+            }
+        });
+
+        size_t workAmount = 0;
+        for (size_t b = 0; b < B; b++) {
+            workAmount += sequenceLengthB[b];
+        }
+
+        // Parallelization could not be made directly by T due to output index depends on merged classes and
+        // blank index, thus could not be shared between threads. Better to divide operation on two steps.
+        // At the first stage find the maximum index. At second stage merge if needed.
+        // Such approach makes parallelization more efficient.
+        auto threadBody = [&](const int ithr, const int nthr) {
+            size_t start(0lu), end(0lu);
+            splitter(workAmount, nthr, ithr, start, end);
+            if (start >= end)
+                return;
+            size_t tStart = 0lu, bStart = 0lu;
+            for (; bStart < B; bStart++) {
+                tStart += sequenceLengthB[bStart];
+                if (tStart >= start) {
+                    tStart = start - (tStart - sequenceLengthB[bStart]);
+                    break;
+                }
+            }
+
+            size_t workCounter = start;
+
+            for (size_t b = bStart; b < B; ++b) {
+                size_t outputIndex = b * T + tStart;
+                const float* probs = probabilities + b * C + BC * tStart;
+                size_t sequenceLength = sequenceLengthB[b];
+
+                for (size_t t = tStart; t < sequenceLength; ++t) {
+                    int maxClassIdx = 0;
+
+                    float maxProb = probs[0];
+                    ++probs;
+
+                    for (int c = 1; c < C; ++c, ++probs) {
+                        if (*probs > maxProb) {
+                            maxClassIdx = c;
+                            maxProb = *probs;
+                        }
+                    }
+                    probs += CB1;
+                    outputSequences[outputIndex++] = static_cast<float>(maxClassIdx);
+
+                    if (++workCounter >= end) {
+                        return;
+                    }
+                }
+                tStart = 0lu;
+            }
+        }; // thread body
+
+        parallel_nt(0, threadBody);
+
+        parallel_for(B, [&](size_t b) {
+            int prev_class_idx = -1;
+            size_t outputIndex = b * T;
+            const size_t sequenceLength = sequenceLengthB[b];
+            float* shiftedOut = outputSequences + b * T;
+            for (size_t t = 0; t < sequenceLength; ++t) {
+                if (*shiftedOut < blankIndex &&
+                        !(mergeRepeated_ && *shiftedOut == prev_class_idx)) {
```

Comment:
As far as I understand in case `mergeRepeated_  == false` we will never step into condition. In that case seems to be more efficient to put whole 'sequenceLength' under `mergeRepeated_` condition.

Reply:
No. In case mergeRepeated_ == false condition become *shiftedOut < blankIndex && true.